{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tsinfer\n",
    "import msprime\n",
    "import tsdate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_sweep_ts(n, Ne, L, rho=1e-8, mu=1e-8):\n",
    "    sweep_model = msprime.SweepGenicSelection(\n",
    "        position=L/2, start_frequency=0.0001, end_frequency=0.9999, s=0.25, dt=1e-6)\n",
    "    models = [sweep_model, msprime.StandardCoalescent()]\n",
    "    ts = msprime.sim_ancestry(\n",
    "        n, model=models, population_size=Ne, sequence_length=L, recombination_rate=rho, random_seed=1234)\n",
    "    return msprime.sim_mutations(ts, rate=mu, random_seed=4321)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mu = 1e-8\n",
    "pop_size = 10_000\n",
    "rho=1e-8\n",
    "sim_ts = make_sweep_ts(300, Ne=pop_size, L=5_000_000, mu=mu)\n",
    "sim_ts.dump(\"true_topology.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tsinfer + tsdate (one iteration of tsinfer, tsdate v 0.2)\n",
    "i_ts = tsinfer.infer(tsinfer.SampleData.from_tree_sequence(sim_ts))\n",
    "s_ts = i_ts.simplify()\n",
    "d_ts = tsdate.date(\n",
    "    s_ts,\n",
    "    mutation_rate=mu, max_shape=1000)\n",
    "d_ts.dump(\"tsinfer_tsdated.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Relate inference\n",
    "import subprocess\n",
    "import tempfile\n",
    "import tskit\n",
    "import numpy as np\n",
    "def ts_to_haps_sample(ts, haps_output, sample_output, chromosome_number=1, sample_name_field=\"name\"):\n",
    "    \"\"\"\n",
    "    Output the tree sequence as in haps / sample format as required by Relate and ARGneedle\n",
    "    (see https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#hapsample)\n",
    "\n",
    "    ``haps_output`` and ``sample_output`` are the filehandles to which the data will be written.\n",
    "    To obtain either as strings, you can pass an io.StringIO object here.\n",
    "\n",
    "    ``sample_name_field`` gives the metadata field in which to look up names to use in the\n",
    "    output sample file. Where possible, names are taken from the associated individual metadata.\n",
    "    If samples are not associated with individuals (i.e. this is haploid data), then\n",
    "    names are taken from node metadata. If no ``sample_name_field`` is present in the metadata,\n",
    "    the names used are \"Individual_N\" if samples are associated with individuals, or \"Sample_N\"\n",
    "    otherwise.\n",
    "\n",
    "    Returns an array of the site_ids that were written to the haps file (sites\n",
    "    with 1 allele or > 2 alleles are skipped)\n",
    "\n",
    "    .. example::\n",
    "\n",
    "        with open(\"out.haps\", \"wt\") as haps, open(\"out.sample\", \"wt\") as sample:\n",
    "            ts_to_haps_sample(ts, haps, sample)\n",
    "    \"\"\"\n",
    "    used = np.zeros(ts.num_sites, dtype=bool)\n",
    "    for v in ts.variants():\n",
    "        if len(v.alleles) == 1:\n",
    "            continue\n",
    "        if len(v.alleles) > 2:\n",
    "            print(f\"Multialleic site ({v.alleles}) at position {v.site.position} ignored\")\n",
    "            continue\n",
    "        used[v.site.id] = True\n",
    "        print(\n",
    "            str(chromosome_number),\n",
    "            f\"SNP{v.site.id}\",\n",
    "            int(v.site.position),\n",
    "            v.alleles[0],\n",
    "            v.alleles[1],\n",
    "            \" \".join([str(g) for g in v.genotypes]),\n",
    "            sep=\" \",\n",
    "            file=haps_output,\n",
    "        )\n",
    "\n",
    "    print(\"ID_1 ID_2 missing\", file=sample_output)\n",
    "    print(\"0    0    0\", file=sample_output)\n",
    "    individuals = ts.nodes_individual[ts.samples()]\n",
    "    if np.all(individuals == tskit.NULL):\n",
    "        # No individuals, just use node metadata\n",
    "        pass\n",
    "    else:\n",
    "        if np.any(individuals == tskit.NULL):\n",
    "            raise ValueError(\"Some samples have no individuals\")\n",
    "        _, counts = np.unique(individuals, return_counts=True)\n",
    "        if np.all(counts == 2):\n",
    "            if np.any(np.diff(individuals)[0::2]) != 0:\n",
    "                ValueError(\"Pairs of adjacent samples must come from the same individual\")\n",
    "        elif np.all(counts == 1):\n",
    "            pass\n",
    "        else:\n",
    "            raise ValueError(\"Must have all diploid or all haploid samples\")\n",
    "    samples = ts.samples()\n",
    "    i=0\n",
    "    while i < len(samples):\n",
    "        ind1 = ts.node(samples[i]).individual\n",
    "        ind2 = tskit.NULL\n",
    "        if ind1 == tskit.NULL:\n",
    "            try:\n",
    "                name = ts.node(samples[i]).metadata[sample_name_field].replace(\" \", \"_\")\n",
    "            except (TypeError, KeyError):\n",
    "                name = f\"Sample_{samples[i]}\"\n",
    "        else:\n",
    "            try:\n",
    "                name = ts.individual(ind1).metadata[sample_name_field].replace(\" \", \"_\")\n",
    "            except (TypeError, KeyError):\n",
    "                name = f\"Individual_{ind1}\"\n",
    "            try:\n",
    "                ind2 = ts.node(samples[i+1]).individual\n",
    "            except IndexError:\n",
    "                pass\n",
    "        if ind2 == tskit.NULL or ind2 != ind1:\n",
    "            print(f'{name} NA 0', file=sample_output)\n",
    "            i += 1\n",
    "        else:\n",
    "            print(f'{name} {name} 0', file=sample_output)\n",
    "            i += 2\n",
    "    return np.where(used)[0].astype(ts.mutations_site.dtype)\n",
    "\n",
    "def run_relate(ts, population_size, mu, rho, random_seed=111, path_to_relate=\"/home/savita/DPhil/code/tsbrowse_paper/Relate/relate_v1.2.2_x86_64_dynamic/\"):\n",
    "    dir = \"/home/savita/DPhil/code/tsbrowse_paper/Relate/\"\n",
    "    prefix = \"test\"\n",
    "    with open(dir + prefix + \".haps\", \"wt\") as haps, open(dir + prefix + \".sample\", \"wt\") as sample:\n",
    "        # ts_to_haps_sample routine from https://github.com/tskit-dev/tsconvert/issues/55#issuecomment-1831959994\n",
    "        ts_to_haps_sample(ts, haps, sample)\n",
    "    with tempfile.NamedTemporaryFile(\"wt\") as temp:\n",
    "        cM_per_MB = rho * 1e8\n",
    "        print(\"pos\", \"COMBINED_rate\", \"Genetic_Map\", sep=\" \", file=temp)\n",
    "        print(0, f\"{cM_per_MB:.5f}\", 0, sep=\" \", file=temp)\n",
    "        print(\n",
    "            int(ts.sequence_length),\n",
    "            f\"{cM_per_MB:.5f}\",\n",
    "            ts.sequence_length / 1e6 * cM_per_MB,\n",
    "            sep=\" \",\n",
    "            file=temp)\n",
    "        temp.flush()\n",
    "\n",
    "        params = [\n",
    "            path_to_relate + \"bin/Relate\",\n",
    "            \"--haps\", prefix+\".haps\",\n",
    "            \"--sample\", prefix+\".sample\",\n",
    "            \"--map\", temp.name,\n",
    "            \"-o\", \"Relate\",\n",
    "            \"--mode\", \"All\",\n",
    "            \"-m\", f\"{mu}\",\n",
    "            \"-N\", f\"{population_size}\",\n",
    "            \"--seed\", f\"{random_seed}\",\n",
    "        ]\n",
    "        print(f\"running `{' '.join(params)}`\")\n",
    "        subprocess.run(params, cwd=dir)\n",
    "\n",
    "    # Convert to tree sequence format\n",
    "    params = [\n",
    "        path_to_relate + \"/bin/RelateFileFormats\",\n",
    "        \"--mode\", \"ConvertToTreeSequence\",\n",
    "        \"-i\", \"out\",\n",
    "        \"-o\", \"relate\",\n",
    "    ]\n",
    "    print(f\"running `{' '.join(params)}`\")\n",
    "    subprocess.run(params, cwd=dir)\n",
    "    return tskit.load(dir + \"relate.trees\")\n",
    "\n",
    "relate_ts = run_relate(sim_ts, pop_size/2, mu, rho)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir = \"/home/savita/DPhil/code/tsbrowse_paper/ARG-Needle/\"\n",
    "prefix = \"true_topology\"cp ../\n",
    "with open(dir + prefix + \".haps\", \"wt\") as haps, open(dir + prefix + \".sample\", \"wt\") as sample:\n",
    "    sites = ts_to_haps_sample(ts, haps, sample)\n",
    "with open(dir + prefix + \".map\", \"wt\") as map, open(dir + prefix + \".demo\", \"wt\") as demo:\n",
    "    # Make the required mapfile (one line per variant)\n",
    "    # https://palamaralab.github.io/software/argneedle/manual/#genetic-map-mapmapgz\n",
    "    # chromosome SNP_name genetic_position_cM physical_position_bp\n",
    "    for s in sites:\n",
    "        pos = sim_ts.site(s).position\n",
    "        print(\"1\", f\"Site{s}\", f\"{pos * rho * 100}\", f\"{pos}\", sep=\"\\t\", file=map)\n",
    "    print(\"\\t\".join([\"0.0\", str(pop_size)]), file=demo)\n",
    "    print(\"\\t\".join([\"5000.0\", str(pop_size)]), file=demo)\n",
    "#at this point I was getting errors about numpy version needing to be downgraded, and it persisted, so I ran the below steps from the terminal\n",
    "#arg_needle --hap_gz test.haps --map test.map --mode sequence --normalize_demography test.demo --out arg_needl\n",
    "#arg2tskit --arg_path arg_needle.argn --ts_path arg_needle.trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SINGER\n",
    "#wget https://github.com/popgenmethods/SINGER/tree/main/releases/singer-0.1.8-beta-linux-x86_64\n",
    "with open(\"/home/savita/DPhil/code/tsbrowse_paper/SINGER/sim_ts.vcf\", \"w\") as vcf_file:\n",
    "    sim_ts.write_vcf(vcf_file)\n",
    "#for the next step I found it easier to run the command from the terminal\n",
    "#./singer_master -m 1e-8 -vcf sim_ts -start 0 -end 4999999 -output singer_out\n",
    "#./convert_to_tskit -i singer_out -o singer -start 0 -end 99"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#to run tsbrowse on these tree sequences:\n",
    "#python -m tsbrowse preprocess <in.trees>\n",
    "#python -m tsbrowse serve --port 8090 <in.tsbrowse>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tsbrowse_selective_sweep_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
