{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import tsinfer\n",
    "import msprime\n",
    "import tsdate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_sweep_ts(n, Ne, L, rho=1e-8, mu=1e-8):\n",
    "    sweep_model = msprime.SweepGenicSelection(\n",
    "        position=L/2, start_frequency=0.0001, end_frequency=0.9999, s=0.25, dt=1e-6)\n",
    "    models = [sweep_model, msprime.StandardCoalescent()]\n",
    "    ts = msprime.sim_ancestry(\n",
    "        n, model=models, population_size=Ne, sequence_length=L, recombination_rate=rho, random_seed=1234)\n",
    "    return msprime.sim_mutations(ts, rate=mu, random_seed=4321)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mu = 1e-8\n",
    "pop_size = 10_000\n",
    "rho=1e-8\n",
    "sim_ts = make_sweep_ts(300, Ne=pop_size, L=5_000_000, mu=mu)\n",
    "sim_ts.dump(\"true_topology.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tsinfer + tsdate (one iteration of tsinfer, tsdate v 0.2)\n",
    "i_ts = tsinfer.infer(tsinfer.SampleData.from_tree_sequence(sim_ts))\n",
    "s_ts = i_ts.simplify()\n",
    "d_ts = tsdate.date(\n",
    "    s_ts,\n",
    "    mutation_rate=mu, max_shape=1000)\n",
    "d_ts.dump(\"tsinfer_tsdated.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multialleic site (('T', 'C', 'A')) at position 1284745.0 ignored\n",
      "Multialleic site (('C', 'G', 'T')) at position 1307324.0 ignored\n",
      "Multialleic site (('C', 'G', 'T')) at position 2859263.0 ignored\n",
      "Multialleic site (('A', 'G', 'C')) at position 3599479.0 ignored\n",
      "Multialleic site (('A', 'C', 'G')) at position 3757839.0 ignored\n",
      "Multialleic site (('C', 'T', 'A')) at position 3892870.0 ignored\n",
      "Multialleic site (('G', 'A', 'T')) at position 3908018.0 ignored\n",
      "Multialleic site (('G', 'A', 'C')) at position 4071401.0 ignored\n",
      "Multialleic site (('C', 'G', 'T')) at position 4214239.0 ignored\n",
      "running `/home/savita/DPhil/code/tsbrowse_paper/Relate/relate_v1.2.2_x86_64_dynamic/bin/Relate --haps test.haps --sample test.sample --map /tmp/tmplbe4sw2i -o out --mode All -m 1e-08 -N 5000.0 --seed 111`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "*********************************************************\n",
      "---------------------------------------------------------\n",
      "Relate\n",
      " * Authors: Leo Speidel, Marie Forest, Sinan Shi, Simon Myers.\n",
      " * Doc:     https://myersgroup.github.io/relate\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Using:\n",
      "  test.haps\n",
      "  test.sample\n",
      "  /tmp/tmplbe4sw2i\n",
      "with mu = 1e-08 and 2Ne = 5000.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Parsing data..\n",
      "Warning: Will use min 0.0086GB of hard disc.\n",
      "CPU Time spent: 0.160038s; Max Memory usage: 5.9e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Read 600 haplotypes with 10659 SNPs per haplotype.\n",
      "Expected minimum memory usage: 2.8Gb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Starting chunk 0 of 0.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Painting sequences...\n",
      "CPU Time spent: 2.881080s; Max Memory usage: 5.9e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Estimating topologies of AncesTrees in sections 0-0...\n",
      "CPU Time spent: 29.191125s; Max Memory usage: 3e+03Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Propagating mutations across AncesTrees...\n",
      "CPU Time spent: 30.627711s; Max Memory usage: 3e+03Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Inferring branch lengths of AncesTrees in sections 0-0...\n",
      "CPU Time spent: 86.344712s; Max Memory usage: 3e+03Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Combining AncesTrees in Sections...\n",
      "CPU Time spent: 86.571288s; Max Memory usage: 3e+03Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Finalizing...\n",
      "Number of not mapping SNPs: 0\n",
      "Number of flipped SNPs    : 0\n",
      "CPU Time spent: 87.055554s; Max Memory usage: 3e+03Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Done.\n",
      "---------------------------------------------------------\n",
      "*********************************************************\n",
      "\n",
      "---------------------------------------------------------\n",
      "Converting anc/mut file format (Relate) to tree sequence file format (tskit)..\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running `/home/savita/DPhil/code/tsbrowse_paper/Relate/relate_v1.2.2_x86_64_dynamic//bin/RelateFileFormats --mode ConvertToTreeSequence -i out -o out`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPU Time spent: 1.148428s; Max Memory usage: 590.228Mb.\n",
      "---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Relate inference\n",
    "import subprocess\n",
    "import tempfile\n",
    "import tskit\n",
    "import numpy as np\n",
    "def ts_to_haps_sample(ts, haps_output, sample_output, chromosome_number=1, sample_name_field=\"name\"):\n",
    "    \"\"\"\n",
    "    Output the tree sequence as in haps / sample format as required by Relate and ARGneedle\n",
    "    (see https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#hapsample)\n",
    "\n",
    "    ``haps_output`` and ``sample_output`` are the filehandles to which the data will be written.\n",
    "    To obtain either as strings, you can pass an io.StringIO object here.\n",
    "\n",
    "    ``sample_name_field`` gives the metadata field in which to look up names to use in the\n",
    "    output sample file. Where possible, names are taken from the associated individual metadata.\n",
    "    If samples are not associated with individuals (i.e. this is haploid data), then\n",
    "    names are taken from node metadata. If no ``sample_name_field`` is present in the metadata,\n",
    "    the names used are \"Individual_N\" if samples are associated with individuals, or \"Sample_N\"\n",
    "    otherwise.\n",
    "\n",
    "    Returns an array of the site_ids that were written to the haps file (sites\n",
    "    with 1 allele or > 2 alleles are skipped)\n",
    "\n",
    "    .. example::\n",
    "\n",
    "        with open(\"out.haps\", \"wt\") as haps, open(\"out.sample\", \"wt\") as sample:\n",
    "            ts_to_haps_sample(ts, haps, sample)\n",
    "    \"\"\"\n",
    "    used = np.zeros(ts.num_sites, dtype=bool)\n",
    "    for v in ts.variants():\n",
    "        if len(v.alleles) == 1:\n",
    "            continue\n",
    "        if len(v.alleles) > 2:\n",
    "            print(f\"Multialleic site ({v.alleles}) at position {v.site.position} ignored\")\n",
    "            continue\n",
    "        used[v.site.id] = True\n",
    "        print(\n",
    "            str(chromosome_number),\n",
    "            f\"SNP{v.site.id}\",\n",
    "            int(v.site.position),\n",
    "            v.alleles[0],\n",
    "            v.alleles[1],\n",
    "            \" \".join([str(g) for g in v.genotypes]),\n",
    "            sep=\" \",\n",
    "            file=haps_output,\n",
    "        )\n",
    "\n",
    "    print(\"ID_1 ID_2 missing\", file=sample_output)\n",
    "    print(\"0    0    0\", file=sample_output)\n",
    "    individuals = ts.nodes_individual[ts.samples()]\n",
    "    if np.all(individuals == tskit.NULL):\n",
    "        # No individuals, just use node metadata\n",
    "        pass\n",
    "    else:\n",
    "        if np.any(individuals == tskit.NULL):\n",
    "            raise ValueError(\"Some samples have no individuals\")\n",
    "        _, counts = np.unique(individuals, return_counts=True)\n",
    "        if np.all(counts == 2):\n",
    "            if np.any(np.diff(individuals)[0::2]) != 0:\n",
    "                ValueError(\"Pairs of adjacent samples must come from the same individual\")\n",
    "        elif np.all(counts == 1):\n",
    "            pass\n",
    "        else:\n",
    "            raise ValueError(\"Must have all diploid or all haploid samples\")\n",
    "    samples = ts.samples()\n",
    "    i=0\n",
    "    while i < len(samples):\n",
    "        ind1 = ts.node(samples[i]).individual\n",
    "        ind2 = tskit.NULL\n",
    "        if ind1 == tskit.NULL:\n",
    "            try:\n",
    "                name = ts.node(samples[i]).metadata[sample_name_field].replace(\" \", \"_\")\n",
    "            except (TypeError, KeyError):\n",
    "                name = f\"Sample_{samples[i]}\"\n",
    "        else:\n",
    "            try:\n",
    "                name = ts.individual(ind1).metadata[sample_name_field].replace(\" \", \"_\")\n",
    "            except (TypeError, KeyError):\n",
    "                name = f\"Individual_{ind1}\"\n",
    "            try:\n",
    "                ind2 = ts.node(samples[i+1]).individual\n",
    "            except IndexError:\n",
    "                pass\n",
    "        if ind2 == tskit.NULL or ind2 != ind1:\n",
    "            print(f'{name} NA 0', file=sample_output)\n",
    "            i += 1\n",
    "        else:\n",
    "            print(f'{name} {name} 0', file=sample_output)\n",
    "            i += 2\n",
    "    return np.where(used)[0].astype(ts.mutations_site.dtype)\n",
    "\n",
    "def run_relate(ts, population_size, mu, rho, random_seed=111, path_to_relate=\"/home/savita/DPhil/code/tsbrowse_paper/Relate/relate_v1.2.2_x86_64_dynamic/\"):\n",
    "    dir = \"/home/savita/DPhil/code/tsbrowse_paper/Relate/\"\n",
    "    prefix = \"test\"\n",
    "    with open(dir + prefix + \".haps\", \"wt\") as haps, open(dir + prefix + \".sample\", \"wt\") as sample:\n",
    "        # ts_to_haps_sample routine from https://github.com/tskit-dev/tsconvert/issues/55#issuecomment-1831959994\n",
    "        ts_to_haps_sample(ts, haps, sample)\n",
    "    with tempfile.NamedTemporaryFile(\"wt\") as temp:\n",
    "        cM_per_MB = rho * 1e8\n",
    "        print(\"pos\", \"COMBINED_rate\", \"Genetic_Map\", sep=\" \", file=temp)\n",
    "        print(0, f\"{cM_per_MB:.5f}\", 0, sep=\" \", file=temp)\n",
    "        print(\n",
    "            int(ts.sequence_length),\n",
    "            f\"{cM_per_MB:.5f}\",\n",
    "            ts.sequence_length / 1e6 * cM_per_MB,\n",
    "            sep=\" \",\n",
    "            file=temp)\n",
    "        temp.flush()\n",
    "\n",
    "        params = [\n",
    "            path_to_relate + \"bin/Relate\",\n",
    "            \"--haps\", prefix+\".haps\",\n",
    "            \"--sample\", prefix+\".sample\",\n",
    "            \"--map\", temp.name,\n",
    "            \"-o\", \"Relate\",\n",
    "            \"--mode\", \"All\",\n",
    "            \"-m\", f\"{mu}\",\n",
    "            \"-N\", f\"{population_size}\",\n",
    "            \"--seed\", f\"{random_seed}\",\n",
    "        ]\n",
    "        print(f\"running `{' '.join(params)}`\")\n",
    "        subprocess.run(params, cwd=dir)\n",
    "\n",
    "    # Convert to tree sequence format\n",
    "    params = [\n",
    "        path_to_relate + \"/bin/RelateFileFormats\",\n",
    "        \"--mode\", \"ConvertToTreeSequence\",\n",
    "        \"-i\", \"out\",\n",
    "        \"-o\", \"out\",\n",
    "    ]\n",
    "    print(f\"running `{' '.join(params)}`\")\n",
    "    subprocess.run(params, cwd=dir)\n",
    "    return tskit.load(dir + \"out.trees\")\n",
    "\n",
    "relate_ts = run_relate(sim_ts, pop_size/2, mu, rho)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.0.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"/home/savita/miniconda3/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/home/savita/miniconda3/lib/python3.10/runpy.py\", line 86, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/savita/miniconda3/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/savita/miniconda3/lib/python3.10/asyncio/base_events.py\", line 1906, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/savita/miniconda3/lib/python3.10/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_15169/2714296053.py\", line 1, in <module>\n",
      "    import arg_needle\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/arg_needle/__init__.py\", line 18, in <module>\n",
      "    from .inference import (\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/arg_needle/inference.py\", line 30, in <module>\n",
      "    from fastcluster import linkage\n",
      "  File \"/home/savita/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/fastcluster.py\", line 37, in <module>\n",
      "    from _fastcluster import linkage_wrap, linkage_vector_wrap\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "numpy.core.multiarray failed to import",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[50], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01marg_needle\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun_argneedle\u001b[39m(ts, population_size, mu, rho):\n",
      "File \u001b[0;32m~/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/arg_needle/__init__.py:18\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This file is part of the ARG-Needle genealogical inference and\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# analysis software suite.\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Copyright (C) 2023 ARG-Needle Developers.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# You should have received a copy of the GNU General Public License\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# along with this program.  If not, see <http://www.gnu.org/licenses/>.\u001b[39;00m\n\u001b[0;32m---> 18\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01minference\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m     19\u001b[0m     build_arg_simulation, build_arg, extend_arg,\n\u001b[1;32m     20\u001b[0m     add_default_arg_building_arguments, normalize_arg, trim_arg\n\u001b[1;32m     21\u001b[0m )\n\u001b[1;32m     23\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbuild_arg_simulation\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbuild_arg\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtrim_arg\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     30\u001b[0m ]\n",
      "File \u001b[0;32m~/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/arg_needle/inference.py:30\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpsutil\u001b[39;00m; process \u001b[38;5;241m=\u001b[39m psutil\u001b[38;5;241m.\u001b[39mProcess(os\u001b[38;5;241m.\u001b[39mgetpid())\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m# Third-party packages\u001b[39;00m\n\u001b[0;32m---> 30\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mfastcluster\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m linkage\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtskit\u001b[39;00m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;66;03m# our packages\u001b[39;00m\n",
      "File \u001b[0;32m~/DPhil/code/tsbrowse_paper/selective_sweep_example/tsbrowse_selective_sweep_env/lib/python3.10/site-packages/fastcluster.py:37\u001b[0m\n\u001b[1;32m     32\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpdist\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     33\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe fastcluster.linkage function cannot process \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     34\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mvector data since the function \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     35\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscipy.spatial.distance.pdist could not be  \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     36\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mimported.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 37\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m_fastcluster\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m linkage_wrap, linkage_vector_wrap\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msingle\u001b[39m(D):\n\u001b[1;32m     40\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m'''Single linkage clustering (alias). See the help on the “linkage”\u001b[39;00m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;124;03mfunction for further information.'''\u001b[39;00m\n",
      "\u001b[0;31mImportError\u001b[0m: numpy.core.multiarray failed to import"
     ]
    }
   ],
   "source": [
    "dir = \"/home/savita/DPhil/code/tsbrowse_paper/ARG-Needle/\"\n",
    "prefix = \"test\"\n",
    "with open(dir + prefix + \".haps\", \"wt\") as haps, open(dir + prefix + \".sample\", \"wt\") as sample:\n",
    "    sites = ts_to_haps_sample(ts, haps, sample)\n",
    "with open(dir + prefix + \".map\", \"wt\") as map, open(dir + prefix + \".demo\", \"wt\") as demo:\n",
    "    # Make the required mapfile (one line per variant)\n",
    "    # https://palamaralab.github.io/software/argneedle/manual/#genetic-map-mapmapgz\n",
    "    # chromosome SNP_name genetic_position_cM physical_position_bp\n",
    "    for s in sites:\n",
    "        pos = sim_ts.site(s).position\n",
    "        print(\"1\", f\"Site{s}\", f\"{pos * rho * 100}\", f\"{pos}\", sep=\"\\t\", file=map)\n",
    "    print(\"\\t\".join([\"0.0\", str(pop_size)]), file=demo)\n",
    "    print(\"\\t\".join([\"5000.0\", str(pop_size)]), file=demo)\n",
    "#at this point I was getting errors about numpy version needing to be downgraded, and it persisted, so I ran the below steps from the terminal\n",
    "#arg_needle --hap_gz test.haps --map test.map --mode sequence --normalize_demography test.demo --out arg_needl\n",
    "#arg2tskit --arg_path arg_needle.argn --ts_path arg_needle.trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>9231</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>5000000.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>600</td></tr>\n",
       "                      <tr><td>Total Size</td><td>2.1 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>33929</td>\n",
       "                      <td>1.0 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>300</td>\n",
       "                      <td>8.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>10681</td>\n",
       "                      <td>386.0 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>7360</td>\n",
       "                      <td>201.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>224 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>1.9 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>10668</td>\n",
       "                      <td>260.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7f3fc9490b50>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SINGER\n",
    "#wget https://github.com/popgenmethods/SINGER/tree/main/releases/singer-0.1.8-beta-linux-x86_64\n",
    "with open(\"/home/savita/DPhil/code/tsbrowse_paper/SINGER/sim_ts.vcf\", \"w\") as vcf_file:\n",
    "    sim_ts.write_vcf(vcf_file)\n",
    "#for the next step I found it easier to run the command from the terminal\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tsbrowse_selective_sweep_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
