%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%

%%%CONTEMPORARY%%%
%\documentclass[unnumsec,webpdf,contemporary,large]{oup-authoring-template}%
\documentclass[unnumsec,webpdf,contemporary,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,contemporary,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,contemporary,small]{oup-authoring-template}

%%%MODERN%%%
%\documentclass[unnumsec,webpdf,modern,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,modern,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,small]{oup-authoring-template}

%%%TRADITIONAL%%%
%\documentclass[unnumsec,webpdf,traditional,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,traditional,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,namedate,webpdf,traditional,medium]{oup-authoring-template}
%\documentclass[namedate,webpdf,traditional,small]{oup-authoring-template}

%\onecolumn % for one column layouts

%\usepackage{showframe}

\graphicspath{{Fig/}}

% line numbers
%\usepackage[mathlines, switch]{lineno}
%\usepackage[right]{lineno}
\usepackage{hyperref}
\usepackage{lmodern}
%\usepackage{graphicx}
%\usepackage{xr}

\begin{document}

\journaltitle{Bioinformatics}
\journaltitle{Preprint}
\DOI{DOI HERE}
\copyrightyear{2024}
% \pubyear{2019}
% \access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}

\firstpage{1}

\title[tsbrowse]{Tsbrowse: a genome browser for Ancestral Recombination Graphs}

\author[1$\ast$]{Savita Karthikeyan}
\author[1]{Ben Jeffery}
\author[1]{Duncan Mbuli-Robertson}
\author[1]{Jerome Kelleher \ORCID{0000-0000-0000-0000}}

\authormark{Karthikeyan et al.}

\address[1]{\orgdiv{Big Data Institute, Li Ka Shing Centre for Health Information and Discovery}, \orgname{University of Oxford}, \orgaddress{\street{Old Road Campus}, \postcode{Oxford OX3 7LF}, \country{United Kingdom}}}


\corresp[$\ast$]{Corresponding author. \href{email:savita.karthikeyan@st-hughs.ox.ac.uk}{savita.karthikeyan@st-hughs.ox.ac.uk}}

\received{Date}{0}{Year}
\revised{Date}{0}{Year}
\accepted{Date}{0}{Year}

\abstract{
%\textbf{Motivation:}
Ancestral Recombination Graphs (ARGs) represent the interwoven paths of genetic
ancestry for a set of recombining sequences. The ability to capture the
evolutionary history of samples makes ARGs valuable in a wide range of
applications in population and statistical genetics. ARG-based approaches are
increasingly becoming a part of genetic data analysis pipelines due to
breakthroughs enabling ARG inference at biobank-scale. However, there is a lack
of visualisation tools, which are crucial for validating inferences and
generating hypotheses. 
%\textbf{Results:} 
We present \texttt{tsbrowse}, an open-source Python web-app for the interactive
visualisation of the fundamental building-blocks of ARGs, i.e.,  nodes, edges
and mutations. We demonstrate the application of \texttt{tsbrowse} to various
data sources and scenarios, and highlight its key features of browsability
along the genome, user interactivity, and scalability to very large sample
sizes.\\ \textbf{Availability:}\\ Python package:
\texttt{\url{https://pypi.org/project/tsbrowse/}}, \\ Development version:
\texttt{\url{https://github.com/tskit.dev/tsbrowse}}, \\ Documentation:
\texttt{\url{https://tskit.dev/tsbrowse/docs/}}\\
%\textbf{Contact:} \href{name@email.com}{name@email.com}\\
%\textbf{Supplementary information:} Available at
%\textit{\textcolor{red}{\href{https://www.google.com}{Bioinformatics}
%online}.}}
}

\keywords{Ancestral Recombination Graphs, Holoviz, biobank-scale ARG
visualisation, genome browser} \maketitle \section{Introduction} Visualisation
is an essential component of data analysis pipelines. In the past few decades,
genome browsers have become synonymous with sequence visualisation, by
providing user-friendly, interactive interfaces to navigate genomic sequences
and annotations. From early standalone tools to manage \textit{C. elegans} data
\citep{Eeckman1995}, to modern web-apps for vertebrate genomes
\citep{Birney2004,Rangwala2024,Nassar2023}, browsers have continued to evolve
with the changing landscape of genomics. Modern genome browsers can load
datasets from a URL or local file system \citep{IGV-web,Lee2013,Robinson2023}.
The data security and scalability offered by this architecture ensures that
genome browsers continue to be relevant in the biobank era.

Biobank-driven research has accelerated the development of several
technologies, including the ability to infer Ancestral Recombination Graphs
(ARGs) from large-scale genetic data
\citep{Kelleher2019,Speidel2019,Zhang2023,Gunnarsson2024.08.31.610248,Deng2024.03.16.585351}.
ARGs encode the interwoven paths of genetic inheritance resulting from
recombination, and can be represented as genetic genealogies along the sequence
\citep{Wong2024,Hudson1983,Griffiths1997}. They capture rich detail about
ancestral processes influencing the samples, and have garnered great interest
in several applications in population and statistical genetics. These include
the detection of ultra-rare genetic variation \citep{Zhang2023}, the mapping of
quantitative trait loci \citep{Link2023}, and the sparse and efficient modeling
of linkage disequilibrium \citep{Nowbandegani2023}. It is evident that ARG
inference is increasingly becoming a part of genetic data analysis pipelines.
Yet, ARG visualisation, which is crucial for downstream analyses remains
challenging due to the complexity of inferences from empirical data. Existing
visualisation tools (see \hyperref[sec:Discussion]{Discussion}) focus on the
semantic representation of ARGs, but such representations have limited utility
in informing downstream analyses. 

To address this gap, we have developed \texttt{tsbrowse}, an open-source app
intended as a genome browser for ARGs. Its  focus is on visualising the
information structure in ARGs, thereby improving  inference and analysis at
biobank-scale. \texttt{Tsbrowse} provides interactive visual summaries of  the
fundamental properties of ARGs, i.e. the mutations, nodes, edges and trees they
encode. \texttt{Tsbrowse} operates on ARGs encoded as succinct tree sequences
\citep{Wong2024}. This is the native format of \texttt{tskit} \citep{tskit}, a
widely-used library for interchanging ARGs inferred using a variety of methods
\citep{Speidel2019,Zhang2023,Haller2019,Baumdicker2021,Gunnarsson2024.08.31.610248,Deng2024.03.16.585351}.
The interface is run on a web browser. It works on biobank-scale datasets
without requiring \textcolor{red}{raw data uploads to a server}. This is ideal
for performing quality control and exploratory analyses on ARGs inferred from
biobank datasets. \section{Results} \subsection{Architecture} \texttt{Tsbrowse}
is run in a step-wise approach (Figure \ref{fig:Figure_1}), which makes it
convenient to handle large datasets. The initial step (\texttt{tsbrowse
preprocess}) prepares the data for visualisation. It can be performed
separately on a large machine, and the output stored for later use. The second
step (\texttt{tsbrowse serve}), which can be run on a smaller machine, creates
and renders visualisations. This step follows a client-server architecture. On
the server side (the machine on which \texttt{tsbrowse serve} is run), the
Holoviz suite of libraries \citep{Holoviz} is leveraged to create
visualisations based on the pre-processed data. This choice of architecture is
prompted by the need to represent properties of large-scale ARGs in a
meaningful and performant way. For example, a visual representation of millions
of mutations in SARS-Cov-2 ARGs \citep{Zhan2023.06.08.544212} could quickly
lead to performance bottlenecks and overplotting. Datashader
\citep{Datashader}, a component Holoviz package, overcomes this challenge by
rasterising huge datasets (see \hyperref[subsec:User_Interface]{User Interface}
for details, and Supplementary Figure \ref{fig:Supplementary_Figure_4} for a
tsbrowse view of SARS-Cov-2 mutations). A web browser client connects to the
server to render and interact with these visualisations. Heavy data processing
is limited to the server side, keeping the client-side interface lightweight.
For large datasets with complex visualisations, this achieves a smoother user
experience.

\begin{figure} \centering
\includegraphics[width=0.95\linewidth]{figures/MainFig1.png}
\caption{\textbf{Overview of tsbrowse architecture. } In tskit, ARGs are
defined by a collection of tables which broadly map to ARG properties. Exemplar
table names are denoted in dark blue and columns that describe the property in
light blue. In the pre-process step, tables are augmented with additional
information computed for each property (yellow). The output from this
pre-processing step is stored as a \texttt{.tsbrowse} file. Next the serve step
renders the visualisation in a web browser by leveraging tools in the Holoviz
ecosystem. Optional annotations are provided as an input file, allowing users
to overlay contextual information about genes or other sequence features. }
\label{fig:Figure_1} \end{figure}

\subsection{Data model} \label{subsec:Data_Model} The foundation of
\texttt{tsbrowse} is the \texttt{tskit} data model. As \texttt{tsbrowse} is
based on this generalised ARG representation \citep{Wong2024} it can be applied
to ARGs from several widely-used inference methods including \texttt{Relate}
\citep{Speidel2019}, \texttt{ARG-Needle} \citep{Zhang2023} and \texttt{SINGER}
\citep{Deng2024.03.16.585351}. In \texttt{tskit}, ARGs are encoded as a
collection of tables, each of which store information about a specific ARG
property. For example, the Node table describes sampled or ancestral sequences
and records their time of birth, and the individual and population they belong
to. The Edge table describes the relationship between pairs of nodes, and
records the node IDs and the genomic locations over which the edge spans. The
Mutation table records mutation events in the history of the sample, including
the change of allelic state, time of event and node in which the event
occurred. The Node and Edge tables describe the genealogical relationships
between the sampled and ancestral sequences, while the Mutation table describes
sequence variation contained in the samples \citep{Ralph2020}. This compact
representation enables the efficient storage and querying of biobank-scale
ARGs. \texttt{Tsbrowse} augments these tables with additional columns to create
a compressed \texttt{.tsbrowse} file, containing all necessary semantic
metadata required for automatic visualisation (see
\hyperref[subsec:User_Interface]{User Interface}). 

\subsection{User Interface}
\label{subsec:User_Interface}

\begin{figure}
    \centering
    \includegraphics[width=0.95\linewidth]{figures/MainFig2.png}
    \caption{\textbf{A view of ARG edges in \texttt{tsbrowse}.}
    A screenshot of the Edges tab for a 5 Mb sequence simulated with selection coefficient, \textit{s = 0.25}, and beneficial allele situated at \textit{x = 2.5 Mb}. The plot on the left represents each edge as a horizontal line along the sequence (X-axis). The time of parent node is shown on the Y-axis. The persistence of edges across various dimensions is shown through histograms on the right: top, distribution of genomic spans; middle, distribution of time spans; bottom, edge area, which is the product of genomic and time spans.} 

    \label{fig:Figure_2}
\end{figure}

\texttt{Tsbrowse}'s interface is built on the \texttt{Holoviz} ecosystem
\citep{Holoviz}, which brings together \texttt{Datashader}, \texttt{Bokeh},
\texttt{Panel}, \texttt{Holoviews} and other libraries to create seamless data
visualisation pipelines even for large datasets. A key challenge with
large-scale data visualisation is to maintain a responsive user interface.
\texttt{Datashader} aggregates data and renders only the necessary data points
at a given zoom level, thereby minimising the computational load on the
browser. A \texttt{Bokeh} plotting back-end provides useful tools and widgets.
The \textit{pan}, \textit{select} and \textit{zoom} tools enable interactivity
with individual data points, allowing users to gain deeper insights from the
data. The \texttt{serve} step accepts an optional input file containing
sequence annotations. This allows users to overlay contextual information about
genes or other sequence features. 



The user interface is presented as a \texttt{Panel} dashboard, with pages to
describe mutations, nodes, edges and trees in the input ARG. Figure 2
demonstrates the user interface using the Edges page as an example. The plot on
the left is an overview of all 33,929 edges (parent-child relationships between
node-pairs; see \hyperref[subsec:Data_Model]{Data Model}) in a simulated ARG
\hyperref[subsec:sweep_simulation]{described in Supplementary Methods}; each
edge is depicted as a horizontal line connecting the genomic coordinates of the
parent and child nodes on the x-axis. The y-axis denotes the age of either
node. A useful feature of the app is the ability to interact with the
visualisations. The tools on the right of the plot allow the user to examine
edges within a sequence/time slice, and query individual edges. The histograms
to the right provide summaries of edge spans over sequence and time.

%The ability to explore ARG properties across the sequence and interact with the data makes the analysis of local genealogical patterns more intuitive.

 Tsbrowse can be applied to versatile scenarios. Figure \ref{fig:Figure_2}
highlights a use-case in visually characterising ARGs. The input ARG is a
selective sweep dataset simulated using \texttt{msprime}
\citep{Baumdicker2021}. The distinct pattern of edges observed in the middle of
the sequence between \textit{y=0} to \textit{y=2.5} is attributable to a burst
in recent coalescent events, showing that tsbrowse captures this characteristic
of a selective sweep. Supplementary Figure \ref{fig:Supplementary_Figure_1} is
another example of characterising ARGs with tsbrowse. The view of edges from
four different inferences of the same dataset (inferred with \texttt{tsinfer} +
\texttt{tsdate} (Panel A), \texttt{Relate} (Panel B), \texttt{ARG-Needle}
(Panel C) and \texttt{SINGER} (Panel D)), reflects qualitative differences
between inference methods. 

We demonstrate a further application of tsbrowse in validating inferences
(Supplementary Figures \ref{fig:Supplementary_Figure_2} and
\ref{fig:Supplementary_Figure_3}). Supplementary Figure
\ref{fig:Supplementary_Figure_2} shows how mutation patterns in a 2D plot of
genomic position and mutation age are helpful in spotting poorly inferred
regions. Excluding regions with low site density from the variant data improves
the quality of inference (Panel B). Supplementary Figure
\ref{fig:Supplementary_Figure_3} shows the genomic spans of ancestral nodes in
an ARG inferred with tsinfer. Population genetics theory predicts that IBD
(Identical By Descent) segments break down over time, resulting in shorter
ancestral segments further back in the history of the samples. At time,
\textit{y \textgreater{0.8}}, the plot reveals a pattern of node spans contrary
to this expectation. It confirms a known algorithmic artefact due to which
disproportionately long old ancestors are generated.

Supplementary Fig. \ref{fig:Supplementary_Figure_4}, a screenshot of the
Mutations page for SARS-CoV-2 ARGs \citep{Zhan2023.06.08.544212} demonstrates
tsbrowse's ability to handle diverse data sources. This plot represents a total
of \textcolor{red}{X million} mutations from an inference of
\textcolor{red}{1.27 million} sequences. After executing the
\texttt{preprocess} step, the page loads in \textcolor{red}{X} seconds on a
system with \textcolor{red}{Y GB RAM}, showing that it is possible to visualise
large-scale datasets in a performant way. Panel B demonstrates a useful
interactive feature; additional information about a mutation event (for
example, the mutational frequency in predefined populations, or the number of
sample nodes carrying the mutation) is displayed on mouse-over of individual
data points.

\section{Discussion} \label{sec:Discussion} Visualisation techniques are
essential for data-driven insights. Specialised tools such as \textit{Treenome
browser} \citep{Kramer2023} tackle the challenge of semantic visualisation of
large phylogenies. Yet capturing the data model underlying ARGs is crucial for
the systematic exploration of local genealogies. \texttt{Tsbrowse} addresses
this open problem by condensing ARG properties into a set of browsable,
interactive visual summaries, thus aiding in ARG-based data exploration and
hypothesis generation.

In current practice, simulated ARGs are used as ground-truth to validate
inferences from empirical data
\citep{Kelleher2019,Zhang2023,Speidel2019,Gunnarsson2024.08.31.610248}.
Sophisticated techniques
\citep{Haller2019,Baumdicker2021,Adrion2020,Lauterbur2023} to simulate ARGs
under complex, realistic demographic models \cite{Anderson-Trocme2023} now
exist. However, models of simulation do not capture \textit{all} the nuances of
population datasets. This results in a clear gap between ARG inference and
application. \texttt{Tsbrowse} captures ARG properties in a model-free way,
providing the means to qualitatively assess inferences. With notable features
of browsability along the genome, scalability and applicability to diverse data
sources, inference and simulation methods, \texttt{tsbrowse} has the potential
to be integrated into biobank-scale data pipelines to improve ARG inference and
analysis.

\section*{Acknowledgments}
\section*{Author contributions}
\section*{Supplementary Data}
Supplementary data are available at \textcolor{red}{Bioinformatics online}.
\section*{Conflict of interest}
No competing interest is declared.
\section*{Funding}
This work was supported by funding from the Biotechnology and Biological
Sciences Research Council (UKRI-BBSRC) [grant number BB/T008784/1],
\textcolor{red}{add Novo funding and funding details for Ben, Duncan, Jerome}.
\bibliographystyle{natbib}
\bibliography{reference}

\clearpage \onecolumn \setcounter{figure}{0}  % Reset the figure counter
\renewcommand{\thefigure}{S\arabic{figure}}  % Change figure numbering to S1,
S2, etc. \section{Supplementary Material} \subsection{Simulation of truth
dataset (Fig. \ref{fig:Figure_2})} \label{subsec:sweep_simulation} Ancestral
histories of 300 samples were simulated with \texttt{SweepGenicSelection}
function in \texttt{msprime (version 1.3.3)}. A combination of models was used:
in the recent past, a selective sweep was simulated with a beneficial allele
situated in the middle of a 5 Mb sequence. The frequency of the allele in the
population was set at 0.0001 at the beginning of the sweep. The allele fixed in
the population at a frequency of 0.9999. The strength of selection was set
using the selection coefficient, \textit{s = 0.25}. A time increment,
\textit{dt = 1e-6} was used to step through the sweep. Mutations were added to
the ARG at a rate of 1e-6. For simulating history before occurrence of the
sweep, a standard coalescent model (Hudson's algorithm \citep{Hudson1983}) was
used until coalescence was achieved. 

\subsection{Inference of selective sweep dataset (Suppl. Fig.
\ref{fig:Supplementary_Figure_1})} The following software were used to infer
ARGs from the truth dataset described above: \texttt{tsinfer version 0.3.3}
\citep{Kelleher2019}, \texttt{tsdate version 0.2.1}, \texttt{Relate version
1.2.2} \cite{Speidel2019}, \texttt{ARG-needle version 1.0.3} \citep{Zhang2023},
\texttt{SINGER version 0.1.8-beta} \citep{Deng2024.03.16.585351}. For all
methods, the following parameters of inference were used: \textit{recombination
rate = 1e-8}, \textit{mutation rate = 1e-8}, \textit{effective population size
= 10,000}. Default values were used for all other inference parameters.

\subsection{Inference of 1000 Genomes dataset (Suppl. Fig.
\ref{fig:Supplementary_Figure_2}, Suppl. Fig.
\ref{fig:Supplementary_Figure_3})} The 1000 Genomes dataset was downloaded from
\textcolor{red}{[insert link]}. Ancestral fasta sequence for chromosome 20
(GRCh38) was downloaded from the Ensembl database. Inference was performed with
a Snakemake pipeline \textcolor{red}{(cite Ben's repo?)} using \texttt{tsinfer
version 0.3.3} for the long arm of chromosome 20 after filtering out duplicate
variant positions, variants with missing or low quality ancestral allele,
singletons, n-1-tons and n-2-tons. Only bi-allelic SNPs were included for
inference. For the inference in Suppl. Fig. 2A, all genomic regions were
included. For the inference in Suppl. Figs 2B and 3, genomic regions with a
site density lesser than 5 per kb were discarded. For Suppl. Fig. 2,
\texttt{tsdate version 0.2.1} was used to estimate the age of ancestral nodes
with \textit{mutation rate = 1.29e-8}, setting all other parameters to default
values. 

\subsection{Inference of SARS-Cov-2 ARGs}

The ARG shown in \ref{fig:Supplementary_Figure_4} was inferred with
\texttt{sc2ts}~\citep{Zhan2023.06.08.544212} using the Viridian
dataset~\citep{hunt2024addressing}. It consists of 516,443 samples, 612,209
nodes, 612,655 edges and 988,303 mutations.

\subsection{} Code to recreate datasets and infer ARGs are available in an
accompanying repository:
\texttt{\url{https://github.com/savitakartik/tsbrowse-paper}}.

\clearpage
\section{Supplementary Figures}
\begin{figure}
    \centering
    \includegraphics[width=0.95\linewidth]{figures/SuppFig1.png}
    \caption{\textbf{\texttt{tsbrowse} applied to inference methods}
    A screenshot of \texttt{tsbrowse}'s Edges view for tsinfer+tsdate (A),
ARG-Needle (B), Relate (C) and SINGER (4) inferences of the truth dataset
simulated under a selective sweep model (in Figure 2 of the main text). For
SINGER, an examplar ARG from the set of output posterior ARG samples is shown.
The X coordinate represents genomic position, each horizontal segment on the
plot shows the genomic coordinates that the edge spans, and Y coordinate shows
time of either the parent or child node in the edge.}
    \label{fig:Supplementary_Figure_1}
\end{figure}

\begin{figure}
    \centering
    % \includegraphics[width=0.95\linewidth]{figures/SuppFig2.png}
    \caption{\textbf{A view of mutations in the 1000G WGS dataset.} A
screenshot of \texttt{tsbrowse}'s Mutations view for a X Mb region on
chromosome 17 inferred from from 3,202 participants from the 1000 Genomes Whole
Genome Sequencing dataset (1000G) \citep{1000G2015, 1000GWGS}. The genomic
position of the mutations is shown on the X axis, and the age of mutation is
represented on the Y axis. Panel B represents an inference after filtering out
low site-density regions in the samples.}
    \label{fig:Supplementary_Figure_2}
\end{figure}

\begin{figure}
    \centering
    \includegraphics[width=0.95\linewidth]{figures/SuppFig3.png}
    \caption{\textbf{A view of the Nodes page for a 1000 Genomes inference.} A
screenshot of \texttt{tsbrowse}'s Nodes view for an inference of the long arm
of chromosome 20 from the 1000 Genomes whole-genome sequencing dataset. At the
top is a plot of node spans over time. The length of sequence that the nodes
span is shown on the X axis, and the time of nodes is shown on the Y axis. The
histogram at the bottom shows the distribution of node spans. }
    \label{fig:Supplementary_Figure_3}
\end{figure}

\begin{figure}
    \centering
    \includegraphics[width=0.95\linewidth]{figures/SuppFig4.png}
    \caption{\textbf{\texttt{tsbrowse} applied to SARS-CoV2 ARGs.}
    A screenshot of \texttt{tsbrowse}'s depiction of 988,303 mutations in an 
    ARG inferred by \texttt{sc2ts}; see  text for details. Also shown are the 
    gene annotations along the X-axis.}
    \label{fig:Supplementary_Figure_4}
\end{figure}
\end{document}
